src/main/java/org/opensearch/knn/index/codec/jvector/JVectorWriter.java

```diff
@@ -526,9 +532,20 @@ public OnHeapGraphIndex getGraph(BuildScoreProvider buildScoreProvider) throws I
             if (mergedFloatVector != null) {
                 log.info("Building graph from merged float vector");
                 var itr = mergedFloatVector.iterator();
-                for (int doc = itr.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = itr.nextDoc()) {
-                    graphIndexBuilder.addGraphNode(doc, randomAccessVectorValues.getVector(doc));
+                // Gather a list of valid document Ids to be streamed later for parallel graph construction
+                final List<Integer> docIds = new ArrayList<>();
+                int doc;
+                while ((doc = itr.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                    docIds.add(doc);
                 }
+
+                // parallel graph construction from the merge documents Ids
+                var vv = randomAccessVectorValues.threadLocalSupplier();
+
+                PhysicalCoreExecutor.pool().submit(() -> {
```

Comment:
We may want to use a custom FJP for this, since it can saturate the CPU.  Something we can change the setting on.

Reply:
Yeah, all thread pools we should provide, added it here just to test whether we get any benefits with timeouts.

Reply:
From what I see we can't really manage it via the OpenSearch shared `ThreadPool` constructs like it does for all other thread pools because it doesn't support a `ForkJoinPool`. In the meantime I created a default static SIMD pool to be used for all index construction related activities within `jVectorFormat`.
I suggest to make a separate follow up in OpenSearch core to make sure we can add FJP pool to the common `ThreadPool` construct to be consistent. Then we can provide it here as well. If that sounds legit I can go ahead and start the issue/discussion.

Reply:
I don't understand why you need to use the opensearch one?  You can just make your own FJP based on the settings.

Reply:
Had an offline discussion with @tjake , the comment refers to being able to provide priority to query over ingestion threads.
The next step in that case will be to create two ForkJoinPools, one for query and one for ingestion.
The ingestion thread pool threads will be prioritized lower.